DNNL: enable conv3d

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40220 DNNL: enable dilation conv
* #35664 DNNL: enable max_pool3d and avg_pool3d
* #35663 DNNL: enable batchnorm3d
* **#35662 DNNL: enable conv3d**

Differential Revision: [D22102408](https://our.internmc.facebook.com/intern/diff/D22102408)

XiaobingSuper committed:
DNNL: enable conv3d

[ghstack-poisoned]

XiaobingSuper commented:
@ngimel , @VitalyFedyunin , this PR is about enable DNNL 3d ops, including conv, pooling and batchnorm, For [resnext3d-101](https://github.com/XiaobingSuper/Resnext3d-for-video-classification), and test on real dataset [UCF101](https://www.crcv.ucf.edu/data/UCF101.php)(input size is 10x3x32x128x170), we can get **~13x** performance improvement compare to native cpu patch on **skx-8180**.  You can see the details in [resnext3d-101](https://github.com/XiaobingSuper/Resnext3d-for-video-classification). Thanks!

XiaobingSuper committed:
Update on "DNNL: enable conv3d"




[ghstack-poisoned]

vincentqb commented:
@VitalyFedyunin -- could you review this PR?

XiaobingSuper committed:
Update on "DNNL: enable conv3d"




[ghstack-poisoned]

XiaobingSuper commented:
@VitalyFedyunin, please help review this code, thanks!

yinghai commented:
@lly-zero-one Could you comment on the perf side?

lly-zero-one commented:
We have few internal changes for the current Conv3d implementation for performance improvement, which will be upstreamed in next week. So I am wondering whether we could do a full performance benchmark. For 2d case, we found the mkldnn conv is 2x slower than the native implementation on a specific production model (I will file a repro).

mingfeima commented:
> We have few internal changes for the current Conv3d implementation for performance improvement, which will be upstreamed in next week. So I am wondering whether we could do a full performance benchmark. For 2d case, we found the mkldnn conv is 2x slower than the native implementation on a specific production model (I will file a repro).

2x performance diff with native implementation is serious... In future if you have similar issues, you may also address this in the **Teams** channel, we will get hands on it asap.
@Jianhui-Li, @uyongw, @jgong5

lly-zero-one commented:
https://github.com/pytorch/pytorch/issues/35937 is for tracking the issue.

VitalyFedyunin commented on aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp:
```diff
@@ -46,25 +46,26 @@ Tensor dense_to_mkldnn(const Tensor& cpu_tensor) {
-    int64_t groups) {
+    int64_t groups,
+    int64_t dims) {
 
```
If it only supports dim == 2 && dim == 3 now, please add TORCH_CHECK

VitalyFedyunin reviewed (changes requested):
This is inconsistent with the approach we use for operators naming, we always explicitly specify 1d,2d,3d operators and we are letting python nn module to dispatch to the proper one. 

Using this approach you are not only will follow convention, but also avoid introducing back incompatible changes. 

This comment applies to all PRs in stack.

XiaobingSuper committed:
Update on "DNNL: enable conv3d"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "DNNL: enable conv3d"




[ghstack-poisoned]

XiaobingSuper commented:
@VitalyFedyunin

XiaobingSuper committed:
Update on "DNNL: enable conv3d"




[ghstack-poisoned]

XiaobingSuper commented:
@ngimel @VitalyFedyunin @albanD

XiaobingSuper committed:
Update on "DNNL: enable conv3d"




[ghstack-poisoned]

XiaobingSuper commented:
@ngimel,  please help merge those PRs, thanks!

fmassa commented on torch/utils/mkldnn.py:
```diff
@@ -104,6 +104,56 @@ def __setstate__(self, state):
         self.bias = state[1].to_mkldnn()
         self.training = state[2]
 
+class MkldnnConv3d(torch.jit.ScriptModule):
```
FYI I believe this is kind of a legacy API as we now compile nn Modules recursively. Cc @eellison

eellison replied:
yea, correct, better to inherit from `torch.nn.Module`, you shouldn't need any other changes

XiaobingSuper replied:
Thanks, I will change it,  and also for other case at next step.

XiaobingSuper replied:
@eellison , if it is inherited from **torch.nn.Module**,  there will has a problem for **torch.jit.save** method, because for a MKLDNN module, the parameters are MKLDNN tensor which are opaque tensors(do not have storage), we will first call **.to_dense()** at  **__getstate__**  to save this script module.  I will changed until this problem can be sovled. thanks!

XiaobingSuper commented on torch/utils/mkldnn.py:
```diff
@@ -104,6 +104,56 @@ def __setstate__(self, state):
+
+    @torch.jit.script_method
+    def forward(self, x):
+        return torch.conv3d(
```
For the first question, DNNL also support dilation for convNd, I can enable it, so there only has one case not supported by DNNL: x is not a float tensor, but for this case,  I think weight is also a float tensor, it will report an error to user when reorder the weight, because it need to call **.to_mkldnn()**first which will check the tensor's type.

For the second question, yes, the name is confused, I will change it.

ngimel replied:
Ok, please enable dilated convolution then. I agree that for correct user inputs this situation should not happen, but even in case of incorrect user inputs (user had float weight when creating a module, but is sending double tensor, or cuda tensor, to forward) the error message should be clear and helpful, and I don't know what will happen in this case.

XiaobingSuper replied:
DNNL dilation conv is enabled now. thanks!

XiaobingSuper committed:
Update on "DNNL: enable conv3d"


Differential Revision: [D22102408](https://our.internmc.facebook.com/intern/diff/D22102408)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -234,8 +234,8 @@ auto ConvParams::use_mkldnn(const at::Tensor& input) const -> bool {
     (input.options().backend() == at::Backend::CPU &&
      input.scalar_type() == kFloat && // only on CPU Float Tensors
      !is_dilated() && // doesn't support dilation
```
should you also remove is_dilated check from here if it's actually supported by mkldnn?

XiaobingSuper replied:
There has another PR to do it, see https://github.com/pytorch/pytorch/pull/40220.

ngimel commented on aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp:
```diff
@@ -86,14 +86,44 @@ Tensor mkldnn_reorder_conv2d_weight(
+    IntArrayRef dilation,
+    int64_t groups) {
+
+  auto stride_vec = expand_param_if_needed(stride, "stride", 3);
```
out of curiosity, why do you need to expand stride etc here, but don't need to do it for 2d conv? If it were not for these expansion calls, reorder_weight functions are exactly the same for 2d and 3d.

XiaobingSuper replied:
Yes, we don't need expand them, they have been expanded at https://github.com/pytorch/pytorch/blob/13bd5992d0f91141322a2a70260648f4d333e87e/torch/nn/modules/conv.py#L402-L405

ngimel commented:
Currently if I send tensor of the wrong type (e.g. double) to mkldnn convolution, I get an error
```
RuntimeError: tensor.scalar_type() == ScalarType::Float INTERNAL ASSERT FAILED at "../aten/src/ATen/native/mkldnn/MKLDNNCommon.cpp":70, please report a bug to PyTorch. itensor_view_from_dense expects float tensor input
```
which is the wrong error type, thrown by TORCH_INTERNAL_ASSERT. Should be TORCH_CHECK instead.

XiaobingSuper committed:
Update on "DNNL: enable conv3d"


Differential Revision: [D22102408](https://our.internmc.facebook.com/intern/diff/D22102408)

[ghstack-poisoned]

XiaobingSuper commented:
> Currently if I send tensor of the wrong type (e.g. double) to mkldnn convolution, I get an error
> 
> ```
> RuntimeError: tensor.scalar_type() == ScalarType::Float INTERNAL ASSERT FAILED at "../aten/src/ATen/native/mkldnn/MKLDNNCommon.cpp":70, please report a bug to PyTorch. itensor_view_from_dense expects float tensor input
> ```
> 
> which is the wrong error type, thrown by TORCH_INTERNAL_ASSERT. Should be TORCH_CHECK instead.

Changed to TORCH_CHECK now.

XiaobingSuper commented:
@VitalyFedyunin

VitalyFedyunin commented:
hi @XiaobingSuper we had to revert this stack (see https://ezyang.github.io/pytorch-ci-hud/build/pytorch-master logs), could you please create new PRs
